### ISSUE TITLE
Tuning SNN Performance

### ISSUE THREAD

**Issue by keeperbitz** [NeuromorphicProcessorProject/snn_toolbox#72]
Hello,

I am working to classify heartbeats with a CNN on Intel Loihi, using SNN-TB. As my data consists of heartbeat signals, a 1D CNN suits this task. However, SNN-TB does have this supported directly so I rewrote my 1D CNN as a 2D CNN. I achieve 98.1% accuracy on my dataset using this model architecture. 

My model architecture is:
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         (None, 187, 1, 1)         0         
conv2d_1 (Conv2D)            (None, 92, 1, 16)         96        
batch_normalization_1 (Batch (None, 92, 1, 16)         64        
conv2d_2 (Conv2D)            (None, 44, 1, 16)         1296      
batch_normalization_2 (Batch (None, 44, 1, 16)         64        
flatten_1 (Flatten)          (None, 704)               0         
dense_1 (Dense)              (None, 16)                11280     
dense_3_mitbih (Dense)       (None, 5)                 85        
Total params: 12,885
Trainable params: 12,821
Non-trainable params: 64

However, after conversion, the SNN on Loihi always predicts the same value. I have tried adjusting the number of timesteps, but this does not change this output.

Essentially I am wondering how I should try tuning/fixing this model to work. Is it possible 1D CNN rewritten as 2D CNN may not be compatible? Should I try some regularization? What are recommendations for improving performance with SNN-TB?

Thanks!

**Comment by rbodo** [comment]
Hi,

I've a working version of 1D conv layers on Loihi in my development branch, will upload and let you know soon.

Unrelated: from the model architecture it seems you might be using BatchNorm layers after the nonlinearity. Just to be clear, the order should always be conv-BN-ReLU, not conv-ReLU-BN, otherwise there will be an accuracy drop when the parser fuses conv and BN.

**Comment by keeperbitz** [comment]
I was able to get a decent acc=86.7% and f1=0.682 using the following structure
1dconv-batchnorm-act-flatten-dropout-dense. So thank you for that!

When scaling to higher # of CNN layers upfront, I find the accuracy degrade. Do you surmise what could cause this degradation in accuracy and how I could investigate improving it?

**Comment by rbodo** [comment]
Nice to hear you got better accuracy this way.

I've uploaded my Conv1D implementation in my fork of the inrc contrib repo [here](https://github.com/rbodo/models). With that you wouldn't need the Conv2D workaround any more. But I guess this switch is only relevant for you at this point if you'd like to use things like causal padding.

Regarding the larger number of layers, it is expected that the accuracy degrades because approximation errors in lower layers get amplified and spread out through the synaptic fanout in higher layers. We've shown in our 2017 Frontiers paper that these errors in principle go away when increasing the simulation duration (though it may take an infeasible amount of timesteps). Another solution would be to use a gradual layer clamping (also described in the paper), where higher layers are fixed for a few time stamps and released once the lower layers have stabilized. You can turn on this behavior on Loihi by setting

```
[Loihi]
clamp_layers = True
interval = 2  # Number of timesteps to freeze each consecutive layer. Would try values between 2 and 10.
```

Other factors that contribute to degradation during conversion are outliers in the weight or bias distribution of the original ANN.  These can be reduced by using a regularization or constraint on the parameters during training. 

Also keep in mind that the weights are quantized to 8 bit when going to Loihi, which may also result in an accuracy drop. You can check that by quantizing the ANN manually and then testing the accuracy. Depending on the result you may want to apply quantization-aware training or post-training quantization.

But before trying any of this, it might be worth checking the spikerates of the SNN and compare them against the ANN activations. You can do that by enabling plots in your config. Then check the log dir and make sure there is a good correlation between spikerates and activations. Sometimes you can tell from certain patterns in the correlation plot if there is something going wrong during conversion. (Running on Loihi with plots enabled can take several times longer than without plots because of the probes. If it takes too long or you run into memory issues, you can probe just a single layer selectively, by using the ``layer_to_probe`` parameter.)

Finally, you could try simulating the converted model in the built-in simulator in the toolbox before going to Loihi. Because it has less constraints, it is usually easier to debug there.

**Comment by wusaifei** [comment]
@rbodo Hi, author. What strategy can the INI emulator use to improve the conversion accuracy of the deep network?For example, you mentioned above that you can improve the accuracy of Loihi emulator conversion by using this one.

`[Loihi]
clamp_layers = True
interval = 2  # Number of timesteps to freeze each consecutive layer. Would try values between 2 and 10.`

What if I'm using the INI emulator?

**Comment by wusaifei** [comment]
@rbodo There is another problem. When softmax_to_relu=True, the accuracy of ANN will decrease, for example, it will decrease from 97% to 94%. What is the reason for this?

**Comment by keeperbitz** [comment]
@rbodo Thanks for the advice. I tried using the fork you referenced with the most recent SNN toolbox version (0.5.0), TF (2.2.0), and Python 3.5.2, but am running into what seems to be a TF error when running the CIFAR-10 tutorial in the nxsdk modules. 

~/snntb_dev/snn_toolbox/snntoolbox/bin/utils.py in run_pipeline(config, queue)
    125                         config.get('paths', 'filename_parsed_model')))
    126 
--> 127         spiking_model.build(parsed_model, **testset)
    128 
    129         # Export network in a format specific to the simulator with which it

~/snntb_dev/snn_toolbox/snntoolbox/simulation/utils.py in build(self, parsed_model, **kwargs)
    440 
    441         print("Compiling spiking model...\n")
--> 442         self.compile()
    443 
    444         # Compute number of operations of ANN.

~/apps-snntb-ncl/nxsdk_modules_ncl/snntoolbox/nx_backend.py in compile(self)
    436             custom_objects = {clampedReLU.__name__: clampedReLU}
    437             self.parsed_model = apply_modifications(self.parsed_model,
--> 438                                                     custom_objects)
    439 
    440         # Set partition environment variable before board is started.

~/apps-snntb-ncl/nxsdk_modules_ncl/snntoolbox/nx_backend.py in apply_modifications(model, custom_objects)
   1534         model.save(model_path)
   1535         return keras.models.load_model(model_path,
-> 1536                                        custom_objects=custom_objects)
   1537     finally:
   1538         os.remove(model_path)

~/pythenv95/lib/python3.5/site-packages/tensorflow/python/keras/saving/save.py in load_model(filepath, custom_objects, compile)
    182     if (h5py is not None and (
    183         isinstance(filepath, h5py.File) or h5py.is_hdf5(filepath))):
--> 184       return hdf5_format.load_model_from_hdf5(filepath, custom_objects, compile)
    185 
    186     if sys.version_info >= (3, 4) and isinstance(filepath, pathlib.Path):

~/pythenv95/lib/python3.5/site-packages/tensorflow/python/keras/saving/hdf5_format.py in load_model_from_hdf5(filepath, custom_objects, compile)
    192       # Compile model.
    193       model.compile(**saving_utils.compile_args_from_training_config(
--> 194           training_config, custom_objects))
    195 
    196       # Set optimizer weights.

~/pythenv95/lib/python3.5/site-packages/tensorflow/python/keras/saving/saving_utils.py in compile_args_from_training_config(training_config, custom_objects)
    219     metrics_config = training_config.get('metrics', None)
    220     if metrics_config is not None:
--> 221       metrics = _deserialize_nested_config(_deserialize_metric, metrics_config)
    222 
    223     # Recover weighted metrics.

~/pythenv95/lib/python3.5/site-packages/tensorflow/python/keras/saving/saving_utils.py in _deserialize_nested_config(deserialize_fn, config)
    260     }
    261   elif isinstance(config, (tuple, list)):
--> 262     return [_deserialize_nested_config(deserialize_fn, obj) for obj in config]
    263 
    264   raise ValueError('Saved configuration not understood.')

~/pythenv95/lib/python3.5/site-packages/tensorflow/python/keras/saving/saving_utils.py in <listcomp>(.0)
    260     }
    261   elif isinstance(config, (tuple, list)):
--> 262     return [_deserialize_nested_config(deserialize_fn, obj) for obj in config]
    263 
    264   raise ValueError('Saved configuration not understood.')

~/pythenv95/lib/python3.5/site-packages/tensorflow/python/keras/saving/saving_utils.py in _deserialize_nested_config(deserialize_fn, config)
    253     return None
    254   if _is_single_object(config):
--> 255     return deserialize_fn(config)
    256   elif isinstance(config, dict):
    257     return {

~/pythenv95/lib/python3.5/site-packages/tensorflow/python/keras/saving/saving_utils.py in _deserialize_metric(metric_config)
    283     # case handling for these in compile, based on model output shape.
    284     return metric_config
--> 285   return metrics_module.deserialize(metric_config)
    286 
    287 

~/pythenv95/lib/python3.5/site-packages/tensorflow/python/keras/metrics.py in deserialize(config, custom_objects)
   3347       module_objects=globals(),
   3348       custom_objects=custom_objects,
-> 3349       printable_module_name='metric function')
   3350 
   3351 

~/pythenv95/lib/python3.5/site-packages/tensorflow/python/keras/utils/generic_utils.py in deserialize_keras_object(identifier, module_objects, custom_objects, printable_module_name)
    390       obj = module_objects.get(object_name)
    391       if obj is None:
--> 392         raise ValueError('Unknown ' + printable_module_name + ':' + object_name)
    393     # Classes passed by name are instantiated with no args, functions are
    394     # returned as-is.

ValueError: Unknown metric function:<lambda>

Have you experienced this before? Does it relate to the comment on line 820? I was curious what versions of SNN-TB, TF, etc. that I should try to run this fork with. For reference, I had been using SNN-TB v0.3.2 for Loihi, and the tutorials/my scripts were working.

Best regards,

KB

**Comment by rbodo** [comment]
> @rbodo Hi, author. What strategy can the INI emulator use to improve the conversion accuracy of the deep network?For example, you mentioned above that you can improve the accuracy of Loihi emulator conversion by using this one.
> 
> ```
> [Loihi]
> clamp_layers = True
> interval = 2  # Number of timesteps to freeze each consecutive layer. Would try values between 2 and 10.
> ```
> 
> What if I'm using the INI emulator?

Look for the `filename_clamp_indices` in the [config](https://snntoolbox.readthedocs.io/en/latest/guide/configuration.html#paths).

**Comment by rbodo** [comment]
> @rbodo There is another problem. When softmax_to_relu=True, the accuracy of ANN will decrease, for example, the accuracy of ANN will decrease from 97% to 94%. What is the reason for this?

It's a different non-linear transformation. Softmax maps all values (including negative ones) to the range (0, 1). ReLU clips negative values to 0. If the output layer contains only negative values, the ReLU will be all zero and the classification most likely be wrong. A softmax would have produced reasonable non-zero values.

**Comment by rbodo** [comment]
> @rbodo Thanks for the advice. I tried using the fork you referenced with the most recent SNN toolbox version (0.5.0), TF (2.2.0), and Python 3.5.2, but am running into what seems to be a TF error when running the CIFAR-10 tutorial in the nxsdk modules.

> ValueError: Unknown metric function:
> 
> Have you experienced this before? It seems to be regarding the custom metrics. I was curious what versions of SNN-TB, TF, etc. that I should try to run this fork with. For reference, I had been using SNN-TB v0.3.2 for Loihi, and the tutorials/my scripts were working.

It's a known issue presumably introduced with a later tensorflow version.

Workarounds
1. Use tensorflow==2.2.0, or
2. Make a small change in the load_model function here:

~/apps-snntb-ncl/nxsdk_modules_ncl/snntoolbox/nx_backend.py in apply_modifications(model, custom_objects)
1535 return keras.models.load_model(model_path, custom_objects=custom_objects)

Pass the argument compile=False. If there's still an error after this change, you may have to compile the loaded model manually by calling
model.compile('sgd', 'categorical_crossentropy', metrics=['accuracy']).

Hope this helps. Will try to submit an update soon.

**Comment by wusaifei** [comment]
Thank you for your reply

**Comment by rbodo** [comment]
The "unknown metric function 'lambda'" error is fixed with commit 720efac. You should be able to use latest tensorflow (2.3.0) together with the development branch of the SNN toolbox.

**Comment by keeperbitz** [comment]
Yes this worked thank you :+1:
